rfcs/25-ui-systems-callbacks.md

```diff
@@ -0,0 +1,493 @@
+These will have similar, but not identical behavior, allowing us to move data from the *components* of the entity into an *event*, commonly of the exact same type.
+In this way, we can use a single system to differentiate behavior.
+
+```rust
+// Operates over any ability buttons we may have in a single system
+fn ability_buttons(mut query: Query<(&mut EventReader<Action>, &mut Timer<Cooldown>, &Ability)>, time: Res<Time>, mut ability_events: EventWriter<Ability>){
+ for actions, cooldown, ability in query.iter_mut(){
+  // Tick down our cooldown timers
+  cooldown.tick(time.delta());
+  
+  for _ in actions {
+   // You can only use abilities that are off cooldown!
+   if cooldown.finished(){
+    // Creates a global ability event using the data for other systems to handle
+    // This should include the originating `Entity` as a field of `Ability`, 
+    // along with other data about its effects
+    ability_events.send(ability);
+   }
+  }
+ }
+}
+```
+
+We can extend this pattern further, with the use of **generic systems**, allowing us to quickly create systems for new types of interactable UI elements.
+
+```rust
+use bevy::prelude::*;
+
+fn main(){
+ App::build()
+  .add_event::<NewPuzzle>()
+  .add_system_to_stage(CoreStage::Input, puzzle_button::<NewPuzzle>.system())
+  .add_event::<ResetPuzzle>()
+  .add_system_to_stage(CoreStage::Input, puzzle_button::<ResetPuzzle>.system())  
+  .add_event::<WriteNumber>()
+  .add_system_to_stage(CoreStage::Input, puzzle_button::<WriteNumber>.system())
+  .run()
+}
+
+// Dataless structs that double as components and events
+struct NewPuzzle;
+struct ResetPuzzle;
+// Also passes the correct number stored in the button into the puzzle game's logic
+struct WriteNumber(u8);
+
+/// Sends the event type associated with the button when pressed
+/// using the data stored on the component of that type
+fn puzzle_button<T: Component + Clone>(
+    query: Query<(&Interaction, &T)>,
+    mut event_writer: EventWriter<T>,
+) {
+    for (interaction, marker) in query.iter() {
+        if *interaction == Interaction::Clicked {
+            event_writer.send(marker.clone());
+        }
+    }
+}
+```
+
+### Specializing behavior
+
+In certain cases though, you may need to have a very large number of UI elements, each with entirely custom behavior.
+Rather than running one (or more!) system per button, you can use a more advanced **callback pattern** using `Callback` components which store a command.
+This command is applied once for each time an `Action` event is received by your entity, taking effect at the end of the current stage.
+
+Under the hood, these are processed by the `callback_system` function found in `CoreStage::Ui`. The `Callback` type and this built-in system are *remarkably* simple:
+
+```rust
+/// `Callback` components are automatically run as commands when 
+enum Callback {
```

Comment:
Per @TheRawMeatball, this should just wrap a `Fn(&mut World)` to simplify things and increase flexibility.